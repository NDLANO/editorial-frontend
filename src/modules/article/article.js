/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { handleActions, createAction } from 'redux-actions';

import { createSelector } from 'reselect';
import { getLocale } from '../locale/locale';
import { convertFieldWithFallback } from '../../util/convertFieldWithFallback';
import formatDate from '../../util/formatDate';

export const fetchArticle = createAction('FETCH_ARTICLE');
export const fetchTopicArticle = createAction('FETCH_TOPIC_ARTICLE');
export const setArticle = createAction('SET_ARTICLE');
export const updateArticle = createAction('UPDATE_ARTICLE');
export const updateArticleSuccess = createAction('UPDATE_ARTICLE_SUCCESS');
export const updateArticleError = createAction('UPDATE_ARTICLE_ERROR');

export const actions = {
  updateArticle,
  fetchArticle,
  setArticle,
  updateArticleSuccess,
  updateArticleError,
  fetchTopicArticle,
};

const initalState = {
  all: {},
  isSaving: false,
};

export default handleActions(
  {
    [setArticle]: {
      next: (state, action) => ({
        ...state,
        all: { ...state.all, [action.payload.id]: { ...action.payload } },
      }),
      throw: state => state,
    },
    [updateArticle]: {
      next: state => ({
        ...state,
        isSaving: true,
      }),
      throw: state => state,
    },
    [updateArticleSuccess]: {
      next: state => ({
        ...state,
        isSaving: false,
      }),
      throw: state => state,
    },
    [updateArticleError]: {
      next: state => ({
        ...state,
        isSaving: false,
      }),
      throw: state => state,
    },
  },
  initalState,
);

const getArticlesFromState = state => state.articles;

export const getArticleById = articleId =>
  createSelector([getArticlesFromState], articles => articles.all[articleId]);

export const getSaving = createSelector(
  [getArticlesFromState],
  articles => articles.isSaving,
);

export const getArticle = articleId =>
  createSelector(
    [getArticleById(articleId), getLocale],
    (article, locale) =>
      article
        ? {
            ...article,
            title: convertFieldWithFallback(article, 'title', ''),
            introduction: convertFieldWithFallback(article, 'introduction', ''),
            visualElement: convertFieldWithFallback(
              article,
              'visualElement',
              {},
            ),
            content: convertFieldWithFallback(article, 'content', ''),
            metaDescription: convertFieldWithFallback(
              article,
              'metaDescription',
              '',
            ),
            tags: convertFieldWithFallback(article, 'tags', []),
            created: formatDate(article.created, locale),
            updated: formatDate(article.updated, locale),
          }
        : undefined,
  );
