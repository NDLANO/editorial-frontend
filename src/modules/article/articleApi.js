/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {
  resolveJsonOrRejectWithError,
  apiResourceUrl,
  fetchWithAccessToken,
} from '../../util/apiHelpers';

const baseUrl = apiResourceUrl('/article-api/v2/articles');

export const fetchArticle = id =>
  fetchWithAccessToken(`${baseUrl}/${id}`).then(resolveJsonOrRejectWithError);

export const fetchTags = () =>
  fetchWithAccessToken(`${baseUrl}/tags/?size=7000`).then(
    resolveJsonOrRejectWithError,
  );

export const fetchLicenses = () =>
  fetchWithAccessToken(`${baseUrl}/licenses`).then(
    resolveJsonOrRejectWithError,
  );

export const updateArticle = article =>
  fetchWithAccessToken(`${baseUrl}/${article.id}`, {
    method: 'PATCH',
    body: JSON.stringify(article),
  }).then(resolveJsonOrRejectWithError);

export const createArticle = article =>
  fetchWithAccessToken(`${baseUrl}/`, {
    method: 'POST',
    body: JSON.stringify(article),
  }).then(resolveJsonOrRejectWithError);

const baseTaxonomyUrl = apiResourceUrl('/taxonomy/v1');

export const fetchTopicArticle = (topicId, locale) =>
  fetchWithAccessToken(
    `${baseTaxonomyUrl}/topics/${topicId}/?language=${locale}`,
  ).then(resolveJsonOrRejectWithError);
