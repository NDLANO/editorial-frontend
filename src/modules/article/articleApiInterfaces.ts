/**
 * Copyright (c) 2020-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ArticleType, Author, AvailabilityType, SearchResultBase } from '../../interfaces';

export type ArticleSearchResult = SearchResultBase<ArticleSearchSummaryApiType>;
export interface ArticleSearchSummaryApiType {
  id: number;
  title: {
    title: string;
    language: string;
  };
  visualElement?: {
    visualElement: string;
    language: string;
  };
  introduction?: {
    introduction: string;
    language: string;
  };
  metaDescription?: {
    metaDescription: string;
    language: string;
  };
  metaImage?: {
    url: string;
    alt: string;
    language: string;
  };
  url: string;
  license: string;
  articleType: string;
  lastUpdated: string;
  supportedLanguages: string[];
  grepCodes: string[];
}

export interface ArticleConverterApiType extends ArticleType {
  availability: AvailabilityType;
  metaData: {
    images: {
      title: string;
      altText: string;
      copyright: Copyright;
      src: string;
      copyText: string;
    }[];
    copyText: string;
  };
}

export interface ArticleApiType {
  id: number;
  oldNdlaUrl?: string;
  revision: number;
  status: {
    status: string;
    other: string[];
  };
  title: {
    title: string;
    language: string;
  };
  content: {
    content: string;
    language: string;
  };
  copyright: Copyright;
  tags: {
    tags: string[];
    language: string;
  };
  requiredLibraries: [
    {
      mediaType: string;
      name: string;
      url: string;
    },
  ];
  visualElement?: {
    visualElement: string;
    language: string;
  };
  metaImage?: {
    url: string;
    alt: string;
    language: string;
  };
  introduction?: {
    introduction: string;
    language: string;
  };
  metaDescription: {
    metaDescription: string;
    language: string;
  };
  created: string;
  updated: string;
  updatedBy: string;
  published: string;
  articleType: string;
  supportedLanguages: string[];
  grepCodes: string[];
  conceptIds: number[];
  availability: string;
}

interface Copyright {
  license: {
    license: string;
    description?: string;
    url?: string;
  };
  origin: string;
  creators: Author[];
  processors: Author[];
  rightsholders: Author[];
  agreementId?: string;
  validFrom?: string;
  validTo?: string;
}
