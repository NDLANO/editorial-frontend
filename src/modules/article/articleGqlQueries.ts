/**
 * Copyright (c) 2023-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { request, gql, Variables } from "graphql-request";
import { useQuery, UseQueryOptions, UseQueryResult } from "@tanstack/react-query";
import config from "../../config";
import { TRANFSFORM_ARTICLE } from "../../queryKeys";
import { apiResourceUrl } from "../../util/apiHelpers";

const gqlEndpoint = config.localConverter
  ? "http://localhost:4000/graphql-api/graphql"
  : apiResourceUrl("/graphql-api/graphql");

interface UseTransformArticle extends Variables {
  content: string;
  language: string;
  visualElement?: string;
  subject?: string;
  previewH5p?: boolean;
  draftConcept?: boolean;
  absoluteUrl?: boolean;
}

export const transformArticleQueryKeys = {
  transformArticle: (params?: Partial<UseTransformArticle>) => [TRANFSFORM_ARTICLE, params] as const,
};

export const usePreviewArticle = (
  content: string,
  language: string,
  visualElement: string | undefined,
  useDraftConcepts: boolean,
  options?: Partial<UseQueryOptions<string>>,
): UseQueryResult<string> => {
  return useTransformArticle(
    {
      content,
      language,
      visualElement,
      previewH5p: true,
      draftConcept: useDraftConcepts,
      absoluteUrl: true,
    },
    options,
  );
};

const transformArticleMutation = gql`
  mutation transformArticle(
    $content: String!
    $visualElement: String
    $subject: String
    $previewH5p: Boolean
    $draftConcept: Boolean
    $absoluteUrl: Boolean
  ) {
    transformArticleContent(
      content: $content
      visualElement: $visualElement
      subject: $subject
      previewH5p: $previewH5p
      draftConcept: $draftConcept
      absoluteUrl: $absoluteUrl
    )
  }
`;

interface ReturnData {
  transformArticleContent: string;
}

export const useTransformArticle = (
  params: UseTransformArticle,
  options?: Partial<UseQueryOptions<string>>,
): UseQueryResult<string> => {
  return useQuery<string>({
    queryKey: transformArticleQueryKeys.transformArticle(params),
    queryFn: async (): Promise<string> => {
      const res = await request<ReturnData, UseTransformArticle>(gqlEndpoint, transformArticleMutation, params, {
        "Accept-Language": params.language,
      });
      return res.transformArticleContent;
    },
    ...options,
  });
};
