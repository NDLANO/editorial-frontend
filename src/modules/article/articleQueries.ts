/**
 * Copyright (c) 2023-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ISearchResultV2 } from '@ndla/types-backend/build/article-api';
import { UseQueryOptions, useQuery } from '@tanstack/react-query';
import { ARTICLE } from '../../queryKeys';
import { ArticleSearchParams, searchArticles } from './articleApi';

const articleSearchQueryKey = (params?: Partial<ArticleSearchParams>) => [ARTICLE, params];

export const useArticleSearch = (
  params: ArticleSearchParams,
  options?: UseQueryOptions<ISearchResultV2>,
) => {
  return useQuery<ISearchResultV2>(
    articleSearchQueryKey(params),
    () => searchArticles(params),
    options,
  );
};
