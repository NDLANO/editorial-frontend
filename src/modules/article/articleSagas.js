/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { take, call, put, select } from 'redux-saga/effects';
import { actions, getArticle } from './article';
import * as api from './articleApi';
import { toEditArticle } from '../../util/routeHelpers';
import * as messageActions from '../../containers/Messages/messagesActions';

export function* fetchArticle(id, language = 'nb') {
  try {
    const tempArticle = yield call(api.fetchArticle, id);
    if (tempArticle.supportedLanguages.includes(language)) {
      const article = yield call(api.fetchArticle, id, language);
      yield put(actions.setArticle({ ...article, language }));

    } else {
      yield put(actions.setArticle({
        id: tempArticle.id,
        language,
        copyright: tempArticle.copyright,
        articleType: 'standard',
        revision: tempArticle.revision,
        supportedLanguages: tempArticle.supportedLanguages, }));
    }
  } catch (error) {
    // TODO: handle error
    console.error(error); //eslint-disable-line
  }
}

export function* watchFetchArticle() {
  while (true) {
    const { payload: { id, language } } = yield take(actions.fetchArticle);
    // console.log('called');
    const article = yield select(getArticle(id));
    if (!article || article.id !== id || article.language !== language) {
      yield call(fetchArticle, id, language);
    }
  }
}

export function* updateArticle(article) {
  try {
    const updatedArticle = yield call(api.updateArticle, article);
    yield put(actions.setArticle(updatedArticle));
    yield put(actions.updateArticleSuccess());
    yield put(messageActions.addMessage({ translationKey: 'form.savedOk' }));
  } catch (error) {
    yield put(actions.updateArticleError());
    // TODO: handle error
    yield put(messageActions.applicationError(error));
  }
}

export function* createArticle(article, history) {
  try {
    const createdArticle = yield call(api.createArticle, article);
    yield put(actions.setArticle(createdArticle));
    history.push(toEditArticle(createdArticle.id, createdArticle.articleType, article.language));
    yield put(actions.updateArticleSuccess());
    yield put(
      messageActions.addMessage({
        translationKey: 'form.createdOk',
      }),
    );
  } catch (error) {
    yield put(actions.updateArticleError());
    // TODO: handle error
    yield put(messageActions.applicationError(error));
  }
}

export function* watchUpdateArticle() {
  while (true) {
    const { payload: { article, history } } = yield take(actions.updateArticle);
    if (article.id) {
      yield call(updateArticle, article);
    } else {
      yield call(createArticle, article, history);
    }
  }
}

export default [watchFetchArticle, watchUpdateArticle];
