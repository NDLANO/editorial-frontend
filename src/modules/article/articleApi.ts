/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { IArticle as DraftApiType } from '@ndla/types-draft-api';
import {
  IArticleV2 as ArticleApiType,
  ISearchResultV2 as ArticleSearchResult,
} from '@ndla/types-article-api';
import {
  resolveJsonOrRejectWithError,
  apiResourceUrl,
  fetchAuthorized,
} from '../../util/apiHelpers';
import config from '../../config';
import { ArticleConverterApiType } from './articleApiInterfaces';
import { LocaleType } from '../../interfaces';

const articleUrl = apiResourceUrl('/article-api/v2/articles');

export const searchArticles = (locale: string, queryString = ''): Promise<ArticleSearchResult> =>
  fetchAuthorized(`${articleUrl}/?language=${locale}&fallback=true${queryString}`).then(r =>
    resolveJsonOrRejectWithError<ArticleSearchResult>(r),
  );

export const searchRelatedArticles = async (
  input: string,
  locale: LocaleType,
  contentType: string,
): Promise<ArticleSearchResult> => {
  await new Promise(resolve => setTimeout(resolve, 50));
  const query = `&type=articles&query=${input}${contentType ? `&content-type=${contentType}` : ''}`;
  return await searchArticles(locale, query);
};

export const getArticle = (id: number, locale: string = 'nb'): Promise<ArticleApiType> =>
  fetchAuthorized(`${articleUrl}/${id}?language=${locale}&fallback=true`).then(r =>
    resolveJsonOrRejectWithError<ArticleApiType>(r),
  );

const articleConverterUrl = config.localConverter
  ? 'http://localhost:3100/article-converter'
  : apiResourceUrl('/article-converter');

export const getArticleFromArticleConverter = (
  id: number,
  locale: string,
): Promise<ArticleConverterApiType> =>
  fetchAuthorized(`${articleConverterUrl}/json/${locale}/${id}`).then(r =>
    resolveJsonOrRejectWithError<ArticleConverterApiType>(r),
  );

export const getPreviewArticle = async (
  article: DraftApiType,
  locale: string,
): Promise<ArticleConverterApiType> => {
  const response = await fetchAuthorized(
    `${articleConverterUrl}/json/${locale}/transform-article?draftConcept=true&previewH5p=true&showVisualElement=true&absoluteUrl=true`,
    {
      headers: {
        'Content-Type': 'application/json',
      },
      method: 'POST',
      body: JSON.stringify({ article }),
    },
  );
  return resolveJsonOrRejectWithError<ArticleConverterApiType>(response);
};
