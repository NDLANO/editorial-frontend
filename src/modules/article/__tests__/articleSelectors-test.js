/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { getArticle, getSaving } from '../article';
import { topicArticle } from './mockArticles';

const state = {
  locale: 'nb',
  articles: {
    isSaving: true,
    all: {
      [topicArticle.id]: topicArticle,
      2: {
        id: '2',
        created: '2014-12-24T10:44:06Z',
        title: [
          { title: 'Tester', language: 'nb' },
          { title: 'Testing', language: 'en' },
        ],
        metaDescription: [
          { metaDescription: 'Beskrivelse', language: 'nb' },
          { metaDescription: 'Description', language: 'en' },
        ],
      },
      3: {
        id: '3',
        created: '2014-11-24T10:44:06Z',
        title: [
          { title: 'Tester', language: 'nb' },
          { title: 'Testing', language: 'en' },
        ],
      },
    },
  },
};

test('articleSelectors getArticle with id', () => {
  expect(getArticle(1)(state).id).toBe('1');
  expect(getArticle(2)(state).id).toBe('2');
  expect(getArticle(3)(state).id).toBe('3');
});

test('articleSelectors getArticle (nb locale)', () => {
  const getArticleSelector = getArticle(1);
  expect(getArticleSelector).toMatchSnapshot();
});

test('articleSelectors getArticle (en locale)', () => {
  const getArticleSelector = getArticle('1');
  const stateWithEnLocale = { ...state, locale: 'en' };
  expect(getArticleSelector(stateWithEnLocale)).toMatchSnapshot();
});

test('articleSelectors getArticle returns undefined if article is not in state', () => {
  const getArticleSelector = getArticle(1337);
  expect(getArticleSelector(state)).toEqual(undefined);
});

test('articleSelectors getSaving', () => {
  expect(getSaving(state)).toBe(true);
});
