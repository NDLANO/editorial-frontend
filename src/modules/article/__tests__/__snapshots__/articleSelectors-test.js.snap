// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articleSelectors getArticle (en locale) 1`] = `
Object {
  "content": "",
  "created": "11/24/2014",
  "id": "3",
  "introduction": "",
  "metaDescription": "",
  "tags": Array [],
  "title": "testing",
  "updated": "Invalid Date",
  "visualElement": Object {},
}
`;

exports[`articleSelectors getArticle (nb locale) 1`] = `[Function]`;
